Replace Respective Image using Custom Component

jasimawan opened the pull request:
I added four new props in the library in **CollageImage.js** and **DynamicCollage.js**. This functionality is available only for Dynamic Colleges as StaticCollage doesn't need it. The four new props is: 

**- onEditButtonPress: PropTypes.func
- EditButtonComponent: PropTypes.func
- editButtonPosition: PropTypes.string
- isEditButtonVisible: PropTypes.bool**

**onEditButtonPress** has a special return statement and it returns matrix used in the collage and index of the picture to be replaced. **onEditButtonPress={(matrix, index) => {}}** I am attaching the code for onEditButtonPress if anyone will face difficulty to update the state of image in the list. 

```
onEditButtonPress={(m, i) => {
            ImagePicker.openPicker({
              multiple: false,
              mediaType: "photo",
            }).then((rep: any) => {
              if (rep.path && rep.path != "") {
                let list = [...images]
                let t = LayoutData[images.length][index].matrix.filter(
                  (v, index, self) => index < m,
                )
                let s = 0
                if (t.length > 0) {
                  s = t.reduce((a, b) => a + b)
                }
                list[s + i] = rep.path // new list of images with replaced respective image
                this.setSate({ images: list })  
                      // update the state of your images data with new list using class based components
                   
                //  OR
                
                setImages(list) // update the state of your images data with new list using hooks
              }
            })
          }
```

**EditButtonComponent** here you can add your custom component or image or icon or anything you want to be displayed on the image. You can style your component as you like. But don't use Button or TouchableOpacity or something like that which have onPress Prop.

**editButtonPosition** this prop is used to set the alignment of EditButtonComponent. I provided the custom styles for this purpose:  **top-left | top-right | bottom-left | bottom-right**.  **This prop is mandatory.**

**isEditButtonVisible** in case you don't want to use replace image functionality set it to **false**. OR If user wants to save his collage image via ViewShot library then you can firstly you can set it to true then when viewshot has been taken change its state to false. So that this button doesn't captured with the collage.


If you want to update anything in this PR. Let me know and i will update it accordingly.
Thanks.

harbolaez commented:
Hey, @jasimawan  thank you for this great! Wanna ask about replacing the image, it seems that you're setting the whole state of images, but this will not re-rerender the child component DynamicCollage? or I'm missing something? I was thinking that you have 3 images and want to replace one, the whole state would change and the state of the other two would get lost?

jasimawan commented:
Hi, @harbolaez thank you for the kind response. No it will only change the image we want to replace, the other images in the list will remain the same. You can see in the example code of **onEditButtonPress**. I made a copy of my images list and replaced the image which i want to be replaced with new image fetched from Image Picker and after that I update the state of images with new updated list. Use my method in your app and you will understand how its updating the the list of images. Thanks.

harbolaez commented:
> Hi, @harbolaez thank you for the kind response. No it will only change the image we want to replace, the other images in the list will remain the same. You can see in the example code of **onEditButtonPress**. I made a copy of my images list and replaced the image which i want to be replaced with new image fetched from Image Picker and after that I update the state of images with new updated list. Use my method in your app and you will understand how its updating the the list of images. Thanks.

I have attached a gyzao on what I'm experiencing https://gyazo.com/928a6ddfbbfd0807da760421e42aca33 -- let me know what you think?

jasimawan commented:
@harbolaez I checked your code and the url you are replacing your image with is not rendering correctly. Test it on some other images url. Secondly there is a problem with your images hook as its not updating the render function. Check your state as well.

lukebrandonfarrell commented:
@jasimawan Thanks for the PR! Appreciate the excitement around this library :) Could you let me know your prettier configuration? then I can push an update on Master and prettify the codebase so we can see more clearly the changes you made.

Also, this code here for updating is crazy: 

```
let list = [...images]
let t = LayoutData[images.length][index].matrix.filter(
  (v, index, self) => index < m,
)
let s = 0
if (t.length > 0) {
  s = t.reduce((a, b) => a + b)
}
list[s + i] = rep.path // new list of images with replaced respective image
```
I wonder if I should add in a replace image API i.e. 

```
collageRef.replaceImage(image, m, i)
```

What do you think? (I can push it as a separate PR)

jasimawan commented:
Yes we can do that as well and I think your suggestion is better than my way of updating images. Glad you respond with a positive feedback and i will definitely enhance it using your suggestion and I will try to implement it this way. And if I feel to be stuck anywhere I will tell you so that you can help me in this regard. 
Thanks.
@lukebrandonfarrell

lukebrandonfarrell commented:
@jasimawan I have prettified the code-base, see: https://github.com/aspect-apps/react-native-images-collage/commit/20e1bb0655f155025edeeb1efc5b682310f5c669

Would you be able to rebase so I can see your changes clearly in this PR?

jasimawan commented:
@lukebrandonfarrell The changes I made you can see them in `CollageImage` file and `DynamicCollage` file. I am attaching the screenshot with it so that you can see it without facing any difficulty. I rebased the code btw.

**CollageImage.js**
![Screen Shot 2021-02-10 at 9 57 38 AM](https://user-images.githubusercontent.com/31315869/107467055-fda26480-6b86-11eb-8fb2-01d87548f2c4.png)

**DynamicCollage.js**
![Screen Shot 2021-02-10 at 9 59 04 AM](https://user-images.githubusercontent.com/31315869/107467067-03984580-6b87-11eb-83d8-770e6df02c88.png)
![Screen Shot 2021-02-10 at 10 00 05 AM](https://user-images.githubusercontent.com/31315869/107467073-05fa9f80-6b87-11eb-949c-55f6ed640d89.png)

lukebrandonfarrell commented:
Thanks @jasimawan ! Can see the changes much more clearly now :) Looks good, will test this on device this evening, then happy to merge and publish

jasimawan commented:
Thanks @lukebrandonfarrell for the kind response and to give me the opportunity to work on this Repo. Waiting for your next response.

lukebrandonfarrell commented on src/DynamicCollage.js:
```diff
@@ -339,6 +343,10 @@ DynamicCollage.propTypes = {
   longPressSensitivity: PropTypes.number, // 1 - 20 - How sensitive is the long press?
+  onEditButtonPress: PropTypes.func,
+  EditButtonComponent: PropTypes.func,
+  editButtonPosition: PropTypes.string,
```
Could we change the prop type to an enum which accepts oneOf: "bottom-right", "bottom-left", "top-right" and "top-left" ?

jasimawan replied:
Yes we can do that, I start working on that and will update the PR. @lukebrandonfarrell

jasimawan replied:
Added enum types in the library and if someone will not enter the right enum type, the app will not crash and the the button will be aligned at top-left by default. @lukebrandonfarrell 
Updating the readme with this change as well and committing.

lukebrandonfarrell commented on src/CollageImage.js:
```diff
@@ -577,6 +578,23 @@ class CollageImage extends React.Component {
+            onPress={this.props.onEditButtonPress}
+            style={{
+              position: "absolute",
+              margin: 20,
```
Could we make the a variable called `editButtonIndent` with a default value of 20? (this will allow people to control the button indentation)

jasimawan replied:
Yes, that would be best if we gave an extra option to the developer. Working on this suggestion. @lukebrandonfarrell

jasimawan replied:
@lukebrandonfarrell  I added the editButtonIntent prop with default value of 20. Updating the pull request.

lukebrandonfarrell reviewed (changes requested):
Nice work on this! Just tested it and looks awesome :) A left a few comments, but after that I am happy to merge and push this as 3.3.0. Thanks again for the PR

lukebrandonfarrell commented:
Oh also @jasimawan if you could update the documentation README.md with these features, that would be of great help! :) Working on the replaceImage API now

jasimawan commented:
I will update the `README.md` file by tomorrow. @lukebrandonfarrell
Thanks for accepting my PR.

lukebrandonfarrell commented:
@all-contributors please add @jasimawan for code and documentation

jasimawan commented:
Updated the pull request with all the suggestions and enhancements. Updated the `README.md` file as well with the latest changes. Requested to review the changes. Kindly check it. @lukebrandonfarrell

lukebrandonfarrell reviewed (approved):
Looks good to me !
